forkmanager/fork_manager.go

```diff
@@ -0,0 +1,222 @@
+	}
+
+	if fork.IsActive {
+		return nil // already activated
+	}
+
+	fork.IsActive = true
+	fork.FromBlockNumber = blockNumber
+
+	for forkHandlerName, forkHandler := range fork.Handlers {
+		fm.addHandler(forkHandlerName, blockNumber, forkHandler)
+	}
+
+	return nil
+}
+
+// DeactivateFork de-activates fork
+// All handlers belong to this fork are also de-activated
+func (fm *forkManager) DeactivateFork(forkName ForkName) error {
+	fm.lock.Lock()
+	defer fm.lock.Unlock()
+
+	fork, exists := fm.forkMap[forkName]
+	if !exists {
+		return fmt.Errorf("fork does not exist: %s", forkName)
+	}
+
+	if !fork.IsActive {
+		return nil // already deactivated
+	}
+
+	fork.IsActive = false
+
+	for forkHandlerName := range fork.Handlers {
+		fm.removeHandler(forkHandlerName, fork.FromBlockNumber)
+	}
+
+	return nil
+}
+
+func (fm *forkManager) GetHandler(name ForkHandlerName, blockNumber uint64) interface{} {
+	fm.lock.Lock()
+	defer fm.lock.Unlock()
+
+	handlers, exists := fm.handlersMap[name]
+	if !exists {
+		return nil
+	}
+
+	// binary search to find first position where handler.FromBlockNumber < blockNumber
+	pos := sort.Search(len(handlers), func(i int) bool {
+		return blockNumber < handlers[i].FromBlockNumber
+	}) - 1
+	if pos < 0 {
+		return nil
+	}
+
+	return handlers[pos].Handler
+}
+
+// IsForkRegistered checks if fork is registered
+func (fm *forkManager) IsForkRegistered(name ForkName) bool {
+	fm.lock.Lock()
+	defer fm.lock.Unlock()
+
+	_, exists := fm.forkMap[name]
+
+	return exists
+}
+
+// IsForkEnabled checks if fork is registered and enabled for specific block
+func (fm *forkManager) IsForkEnabled(name ForkName, blockNumber uint64) bool {
+	fm.lock.Lock()
+	defer fm.lock.Unlock()
+
+	fork, exists := fm.forkMap[name]
+	if !exists {
+		return false
+	}
+
+	return fork.IsActive && fork.FromBlockNumber <= blockNumber
+}
+
+// GetForkBlock returns fork block if fork is registered and activated
+func (fm *forkManager) GetForkBlock(name ForkName) (uint64, error) {
+	fm.lock.Lock()
+	defer fm.lock.Unlock()
+
+	fork, exists := fm.forkMap[name]
+	if !exists {
+		return 0, fmt.Errorf("fork does not exists: %s", name)
+	}
+
+	if !fork.IsActive {
+		return 0, fmt.Errorf("fork is not active: %s", name)
+	}
+
+	return fork.FromBlockNumber, nil
+}
+
+func (fm *forkManager) addHandler(handlerName ForkHandlerName, blockNumber uint64, handler interface{}) {
+	if handlers, exists := fm.handlersMap[handlerName]; !exists {
+		fm.handlersMap[handlerName] = []ForkActiveHandler{
+			{
+				FromBlockNumber: blockNumber,
+				Handler:         handler,
+			},
+		}
+	} else {
+		// keep everything in sorted order
+		index := sort.Search(len(handlers), func(i int) bool {
+			return handlers[i].FromBlockNumber >= blockNumber
+		})
+		handlers = append(handlers, ForkActiveHandler{})
+		copy(handlers[index+1:], handlers[index:])
+		handlers[index] = ForkActiveHandler{
+			FromBlockNumber: blockNumber,
+			Handler:         handler,
+		}
```

Comment:
Should we return an error if somebody tries to register handler for the block number that is already present in the `handlers` slice?

Reply:
this is internal method. If someone calls RegisterHandler with existing fork and handler name, old handler will be overwritten